docs/kubernetes/user-guides/driver-install.md

```diff
@@ -20,4 +20,21 @@ $ ./deploy/setup-project.sh
 $ GCE_PD_SA_DIR=/my/safe/credentials/directory    # Directory to get the service account key
 $ GCE_PD_DRIVER_VERSION=stable                    # Driver version to deploy
 $ ./deploy/kubernetes/deploy-driver.sh
-```
\ No newline at end of file
+```
+
+## Disabling particular CSI driver services
+
+Traditionally, you run the CSI controllers with the GCE PD driver in the same Kubernetes cluster.
+Though, there may be cases where you will only want to run a subset of the available driver services (for example, one scenario is running the controllers outside of the cluster they are serving (while the GCE PD driver still runs inside the served cluster), but there might be others scenarios).
+The CSI driver consists out of these services:
+
+* The **controller** service starts the GRPC server that is used for communication between the CSI components. It is depending on the GCP service account credentials.
+* The **identity** service is responsible to provide identity services like capability information of the CSI plugin.
+* The **node** service implements the various operations for volumes like creation, deletion, etc. It is depending on the GCE metadata service.
+
+The CSI driver has three command line flags, `--run-controller-service`, `--run-identity-service`, and `--run-node-service` which all default to `true`.
+You can disable individually services by setting the respective flags to `false`.
+
+Note: If you want to run the CSI controllers outside of the cluster you might need to specify the `local-zone` parameter in the GCE cloud provider config.
```

Comment:
I now see why this was included in the PR. I think this is actually the main part that aligns with your goal right? Don't we need the `project` as well then? I don't think `local-zone` and `local-project` are the right names for these. They're actually `client-project`, `client-zone`?

Reply:
it seems to me what you're actually looking for is to do a "`MetadataServer` override" in some sense. You could then have a flag that is `use-local-metadata`. When this value is `true (default)` the components will use the metadata service to get local zone and project. When false, a project and zone must be specified through the cloud config and the driver can be deployed outside of the client project.

Reply:
I'm actually not sure why you need a "local-zone" parameter. With the topology feature, the controller can already provision in zones that it's not running in. Is this for the "fallback" behavior when you're not using topology?

Reply:
and the downside of supporting only local-zone is you won't be able to support multi-zone clusters in the future, and potential Regional PD as well.

Reply:
right this would only be for the "default" when no topology is specified. For regional PD if no topology is specified we actually take a specified zone (from metadata service rn) and enumerate all zones in that region to pick a second zone from (of course this doesn't actually work well in practice as you would probably care what other zone you pick). 

@msau42 is actually right, using cloud-config and providing the project + topology should already be enough for the driver to provision and attach volumes in projects/zones it's not running in.

Reply:
Thanks for your comments, though, I'm a little bit confused now and need more help/guidance.

@davidz627 you are right, what I effectively want to prevent is that the driver is talking to the metadata service. For the project name it was already possible to specify it via the cloud config, and the metadata svc was not contacted: https://github.com/kubernetes-sigs/gcp-compute-persistent-disk-csi-driver/blob/master/pkg/gce-cloud-provider/compute/gce.go#L192-L205
I just "replicated" the same approach for the zone (FYI: the name `local-zone` is inspired from https://github.com/kubernetes/legacy-cloud-providers/blob/master/gce/gce.go#L182-L185, but we could call it differently, sure).

> and the downside of supporting only local-zone is you won't be able to support multi-zone clusters in the future, and potential Regional PD as well.

This comment I don't understand - why do you [query the zone from the metadata service](https://github.com/kubernetes-sigs/gcp-compute-persistent-disk-csi-driver/blob/master/pkg/gce-cloud-provider/compute/gce.go#L187-L190) then in the first place (if it's not needed/prevents multi-zone clusters and potential regional PDs)?

From what I see in the code, the zone information is only used to [derive the region name](https://github.com/kubernetes-sigs/gcp-compute-persistent-disk-csi-driver/blob/master/pkg/gce-cloud-provider/compute/gce-compute.go#L87) and to [discover the other zones in this region](https://github.com/kubernetes-sigs/gcp-compute-persistent-disk-csi-driver/blob/master/pkg/gce-cloud-provider/compute/gce-compute.go#L135-L151). There is another usage when [listing disks in a zone](https://github.com/kubernetes-sigs/gcp-compute-persistent-disk-csi-driver/blob/master/pkg/gce-cloud-provider/compute/gce-compute.go#L66-L81) - however, how does all of this work for multi-zone clusters already with today's `master` branch (where the zone is coming from the metadata svc, i.e., the controller can only handle the same zone where it is running in)?

> right this would only be for the "default" when no topology is specified. For regional PD if no topology is specified we actually take a specified zone (from metadata service rn) and enumerate all zones in that region to pick a second zone from (of course this doesn't actually work well in practice as you would probably care what other zone you pick).

In my case, I don't want to assume the controller is running on a GCE instance and has access to the metadata service.

> @msau42 is actually right, using cloud-config and providing the project + topology should already be enough for the driver to provision and attach volumes in projects/zones it's not running in.

What do you more concretely suggest to prevent this unconditional metadata svc call: https://github.com/kubernetes-sigs/gcp-compute-persistent-disk-csi-driver/blob/master/pkg/gce-cloud-provider/compute/gce.go#L187-L190?

Reply:
I think my concrete suggestions going forward would be:

1) Create a flag for `zone` that mirrors the one for `project` right now in the cloud config that will allow you to bypass the metadata config
    - If both project/zone are specified we should be able to turn off the metadata interrogation completely (through some reasonable combination of flags with validation)
2) Don't do any "controller" or "node" disabling in this PR - if there is a follow-on need for that please submit a seperate PR it is fundamentally a different issue.
3) Once you have the ability to turn off the metadata service - you should provide project/zone in the cloud config - however, I would still recommend using the Topology feature to influence volume topology instead of relying on the defaulting behavior (though the defaults may work in your case with single zonal volumes).

Please let me know if any of that doesn't make sense. Hoping it helps

Reply:
Thank you, I'll try going in this direction. Though, I'm not fully aware of the volume topology that you are referring to (as I'm new to CSI). Can you point me in the right direction here, e.g., some place in the code where I can see it in action?

Reply:
Sure, thanks for being so flexible! For topology you can try reading the docs here: https://kubernetes.io/docs/concepts/storage/storage-classes/#volume-binding-mode. It's essentially the way you can influence where PD's are created and therefore scheduled (either dynamically based on Pod using the volume with `WaitForFirstConsumer` or by specifying Topology directly).